### ISSUE TITLE
Layer name is not updated in the legend (layout) after changing the layer name in the layer tree (main window)

### ISSUE THREAD

**Issue by TurboGraphxBeige** [qgis/QGIS#40824]
After changing the name of a layer in the layer tree in the main window, the legend item label for this layer still displays the old name. After changing the layer name in the layer tree then updating the map preview in the layout, the legend properties are updated showing the new name of the layer (unless manually changing "Label" before). However, this change is not drawn on the legend item in the layout. An action firing a redraw for the legend is then needed like:

- Toggling the checkbox "Auto update" or clicking "Update all" (but with the annoyance of re-adding all the map layers and "resetting" the legend)
- Manually changing the "Label" property of the legend item (after manually changing this property, any subsequent layer name change in the layer tree doesn't update the layer name in the legend properties anymore)

I would expect the legend to be drawn with the new layer name after changing its name in the layer tree and updating the map preview in the layout. Specially because the "Label" property is correctly updated (unless a legend item label has been manually updated before).

Tested on QGIS 3.10 (Linux) and on current master branch (Linux).

![Peek 2021-01-02 18-05](https://user-images.githubusercontent.com/59714546/103469014-75ca6e80-4d2d-11eb-9df6-ce2156d49e0b.gif)

![Peek 2021-01-02 18-06](https://user-images.githubusercontent.com/59714546/103468947-7f9fa200-4d2c-11eb-96d7-651095e82f3c.gif)

**Comment by TurboGraphxBeige** [comment]
I'm pretty sure it used to redraw the legend when clicking on it on previous QGIS versions. Regression?

**Comment by gioman** [comment]
> I'm pretty sure it used to redraw the legend when clicking on it on previous QGIS versions

@TurboGraphxBeige if give me an idea on what version it used to work I can give it a try.

**Comment by TurboGraphxBeige** [comment]
It's more or less helpful since the range is quite large, but it's between 3.0 and 3.10. And I'm not even 100% sure the behavior existed!

**Comment by TurboGraphxBeige** [comment]
Nevertheless, I think it would be a good thing to trigger a redraw on a focus on the legend widget. Labels have a similar behavior and we can easily see it when it contains a dynamic expression. Simply focusing on the label will update it with the current values. Though I don't know exactly on which signal the redraw is triggered. Seems to be on focus.

**Comment by TurboGraphxBeige** [comment]
Hey I just discovered that it may probably be an OS specific issue.

I tested with 3.16.2 on Windows 10 and the behavior was normal. After changing the name of a layer in the layer panel, it was also updated in the layout legend (after hitting the "Update Map Preview" button in layout). 

The issue exists on Linux in 3.10, 3.16.3 and Master (prior to https://github.com/qgis/QGIS/pull/41092) all running on Ubuntu 20.04 if that matters.

So the merged PR https://github.com/qgis/QGIS/pull/41092 does solve the issue existing on Linux, but may not be 100% the best solution considering the issue isn't present on Windows? I don't know. Just wanted to let you guys know.
